File: ruby/lib/task_master/ascii_table.rb
module TaskMaster
  class AsciiTable
    def initialize(headers, rows)
      @headers = headers
      @rows = rows
      @column_widths = calculate_column_widths
    end

    def draw
      puts horizontal_line
      print_row(@headers)
      puts horizontal_line
      @rows.each { |row| print_row(row) }
      puts horizontal_line
    end

    private

    def calculate_column_widths
      @headers.map.with_index do |header, i|
        [header, *@rows.map { |row| row[i] }].max_by(&:length).length + 2
      end
    end

    def horizontal_line
      '+' + @column_widths.map { |width| '-' * width }.join('+') + '+'
    end

    def print_row(row)
      row.each_with_index do |cell, i|
        print "| #{cell.ljust(@column_widths[i] - 2)} "
      end
      puts '|'
    end
  end
end

File: javascript/lib/AsciiTable.js
class AsciiTable {
	constructor(headers, rows) {
		this.headers = headers
		this.rows = rows
		this.columnWidths = this.#calculateColumnWidths()
	}

	draw() {
		console.log(this.#horizontalLine())
		this.#printRow(this.headers)
		console.log(this.#horizontalLine())
		this.rows.forEach((row) => { this.#printRow(row) })
		console.log(this.#horizontalLine())
	}

	#calculateColumnWidths() {
		return this.headers.map((header, i) => {
			return [
				header,
				...this.rows.map((row) => {
					return row[i]
				})
			].reduce((max, row) => {
				const currentLength = row.length + 2;
				return Math.max(max, currentLength);
			}, 0);
		})
	}

	#horizontalLine() {
		return "+" + this.columnWidths.map((width) => {
			return "-".repeat(width)
		}).join("+") + "+"
	}

	#printRow(row) {
		row.forEach((cell, i) => {
			const paddedCell = this.#ljust(cell, (this.columnWidths[i] - 2))
		  process.stdout.write(`| ${paddedCell} `)
		})
		console.log("|")
	}

	#ljust(str, length, char = " ") {
		if (str.length >= length) {
			return str;
		}

		const padding = char.repeat(length - str.length)
		return str + padding
	}
}

module.exports = AsciiTable
